## Chapter 10 - Lists

Like a string, a **list** is a sequence of values. In a string, the values are characters; in a list, they can be any type. The values in a list are called **elements** or sometimes items.

There are several ways to create a new list; the simplest is to enclose the elements in square brackets ([ and ]):

In [1]:
[20,30,40]
["hello", 'neighbor', 'have', 'a' , 'ganseett']

['hello', 'neighbor', 'have', 'a', 'ganseett']

A list within another list is **nested**.

A list that contains no elements is called an **empty list**; you can create one with empty brackets, [].

As you might expect, you can assign list values to variables:

In [2]:
rappers = ['Ye', 'Jay', 'LD']

The syntax for accessing the elements of a list is the same as for accessing the characters of a string the bracket operator. The expression inside the brackets specifies the index.

In [3]:
rappers[0]

'Ye'

Unlike strings, lists are mutable. When the bracket operator appears on the left side of an assignment, it identifies the element of the list that will be assigned.

In [4]:
rappers[2] = 'Bino'

In [5]:
rappers

['Ye', 'Jay', 'Bino']

The most common way to traverse the elements of a list is with a for loop. The syntax is the same as for strings:

In [6]:
for rapper in rappers:
    print(rapper)

Ye
Jay
Bino


This works well if you only need to read the elements of the list. But if you want to write or update the elements, you need the indices. A common way to do that is to combine the built-in functions range and len:

In [7]:
numbers = [23,24,12,33]

for i in range(0,len(numbers)):
    numbers[i] = numbers[i] * 2
    
numbers

[46, 48, 24, 66]

Although a list can contain another list, the nested list still counts as a single element.

In [8]:
nested = [['robin', 'blue jay'], ['quail','swan','goose'], ['duck', 'chicken'],['helmet', 'lang', 'visvim']]
len(nested)

4

The + operator concatenates lists and the * operator repeats a list a given number of times:

In [9]:
a = [1, 2, 3]
b = [4, 5, 6]
print(a + b)
print([0] * 4)
print([1, 2, 3] * 3)

[1, 2, 3, 4, 5, 6]
[0, 0, 0, 0]
[1, 2, 3, 1, 2, 3, 1, 2, 3]


The slice operator also works on lists:

In [10]:
rappers[:2]

['Ye', 'Jay']

If you omit the first index, the slice starts at the beginning. If you omit the second, the slice goes to the end. So if you omit both, the slice is a copy of the whole list.

In [11]:
rappers[:]

['Ye', 'Jay', 'Bino']

Since lists are mutable, it is often useful to make a copy before performing operations that modify lists.

A slice operator on the left side of an assignment can update multiple elements:

In [12]:
rappers[0:2] = ['Doom', 'Kendrick']

In [13]:
rappers

['Doom', 'Kendrick', 'Bino']

Python provides methods that operate on lists. For example, **append** adds a new element to the end of a list:

In [14]:
rappers.append('Vince')

In [15]:
rappers

['Doom', 'Kendrick', 'Bino', 'Vince']

**extend** takes a list as an argument and appends all of the elements:

In [16]:
rappers2 = ['Ye', 'Jay', 'Ja', 'BIG']

rappers.extend(rappers2)

rappers

['Doom', 'Kendrick', 'Bino', 'Vince', 'Ye', 'Jay', 'Ja', 'BIG']

**sort** arranges the elements of the list from low to high:

In [17]:
rappers.sort()

rappers

['BIG', 'Bino', 'Doom', 'Ja', 'Jay', 'Kendrick', 'Vince', 'Ye']

Adding up the elements of a list is such a common operation that Python provides it as a built-in function, **sum**:

In [18]:
sum(numbers)

184

An operation like this that combines a sequence of elements into a single value is sometimes called **reduce**.

An operation like capitalize_all is sometimes called a **map** because it “maps” a function (in this case the method capitalize) onto each of the elements in a sequence.

In [19]:
def capitalize_all(t):
    res = []
    for s in t:
        res.append(s.capitalize())
    return res

capitalize_all(rappers)

['Big', 'Bino', 'Doom', 'Ja', 'Jay', 'Kendrick', 'Vince', 'Ye']

An operation like only_upper is called a **filter** because it selects some of the elements and filters out the others.

In [20]:
def only_upper(t):
    res = []
    for s in t:
        if s.isupper():
            res.append(s)
    return res

killas = ['CAM', 'jim jones', 'Santana']

only_upper(killas)

['CAM']

There are several ways to delete elements from a list. If you know the index of the element you want, you can use **pop.**

pop modifies the list and returns the element that was removed. If you don’t provide an index, it deletes and returns the last element.

In [21]:
JJ = killas.pop(1)

print(killas)
print(JJ)

['CAM', 'Santana']
jim jones


If you don’t need the removed value, you can use the **del** operator:

In [22]:
print(rappers)
del rappers[1]
print(rappers)

['BIG', 'Bino', 'Doom', 'Ja', 'Jay', 'Kendrick', 'Vince', 'Ye']
['BIG', 'Doom', 'Ja', 'Jay', 'Kendrick', 'Vince', 'Ye']


If you know the element you want to remove (but not the index), you can use **remove**:

In [23]:
print(rappers)
rappers.remove('Ja')
print(rappers)

['BIG', 'Doom', 'Ja', 'Jay', 'Kendrick', 'Vince', 'Ye']
['BIG', 'Doom', 'Jay', 'Kendrick', 'Vince', 'Ye']


To remove more than one element, you can use del with a slice index:

In [24]:
del rappers[1:3]

print(rappers)

['BIG', 'Kendrick', 'Vince', 'Ye']


The list function breaks a string into individual letters. If you want to break a string into words, you can use the **split** method.

An optional argument called a **delimiter** specifies which characters to use as word boundaries. The second example uses a hyphen as a delimiter.

**join** is the inverse of split. It takes a list of strings and concatenates the elements. join is a string method, so you have to invoke it on the delimiter and pass the list as a parameter.

In [25]:
s = 'pining for the fjords'
t = s.split()
print(t)

spam = 'spam-spam-spam'
delimiter = '-'
tspam = spam.split(delimiter)
print(tspam)

delimiter = ' '
s = delimiter.join(t)
print(s)

['pining', 'for', 'the', 'fjords']
['spam', 'spam', 'spam']
pining for the fjords


If we run these assignment statements:

In [26]:
a = 'banana'
b = 'banana'

We know that a and b both refer to a string, but we don’t know whether they refer to the same string.

To check whether two variables refer to the same object, you can use the is operator.

In [27]:
a is b

True

In this example, Python only created one string object, and both a and b refer to it. But when you create two lists, you get two objects:

In [28]:
a = [1, 2, 3]
b = [1, 2, 3]
a is b

False

In this case we would say that the two lists are **equivalent**, because they have the same elements, but not **identical**, because they are not the same object.

If *a* refers to an object and you assign *b = a*, then both variables refer to the same **object**:

In [29]:
a = [1, 2, 3]
b = a
b is a

True

The association of a variable with an object is called a **reference**. In this example, there are two references to the same object.


An object with more than one reference has *more than one name*, so we say that the object is **aliased**.


If the aliased object is mutable, changes made with one alias affect the other:

In [30]:
print(a)
b[0] = 42
print(a)

[1, 2, 3]
[42, 2, 3]


Although this behavior can be useful, it is error-prone. In general, it is **safer to avoid aliasing** when you are working with mutable objects.

When you pass a list to a function, the function gets a reference to the list. If the function modifies the list, the caller sees the change.

In [31]:
def delete_head(t):
    del t[0]
    
letters = ['a', 'b', 'c']
delete_head(letters)
letters #even though delete_head isn't returning anything, it is still modifying the list

['b', 'c']

The parameter t and the variable letters are **aliases for the same object**.

It is important to distinguish between operations that modify lists and operations that create new lists. For example, the append method modifies a list, but the + operator creates a new list.

This difference is important when you write functions that are supposed to modify lists. For example, this function does *not* delete the head of a list:

In [32]:
def bad_delete_head(t):
    t = t[1:] # WRONG!
    
letters = ['a', 'b', 'c']
bad_delete_head(letters)
letters

['a', 'b', 'c']

The slice operator creates a new list and the assignment makes t refer to it, but that doesn’t affect the caller.

At the beginning of bad_delete_head, *t* and *t4* refer to the same list. At the end, *t* refers to a new list, but *t4* still refers to the original, unmodified list.


An alternative is to write a function that creates and returns a new list. For example, tail returns all but the first element of a list:

In [33]:
def tail(t):
    return t[1:]

letters = ['a', 'b', 'c']
letters_tail = tail(letters)
print(letters)
print(letters_tail)

['a', 'b', 'c']
['b', 'c']


This function leaves the original list unmodified.

**Debugging**

Here
are some common pitfalls and ways to avoid them:

#### 1. Most list methods modify the argument and return None. This is the opposite of the string methods, which return a new string and leave the original alone.

#### 2. Pick an idiom and stick with it.
Part of the problem with lists is that there are too many ways to do things. For example, to remove an element from a list, you can use pop, remove, del, or even a slice assignment.

To add an element, you can use the append method or the + operator.

In [34]:
t=[1,2,3]
x=4
print('These are correct:')
t.append(x)
print(t)
t=[1,2,3]
t = t + [x]
print(t)
t=[1,2,3]
t += [x]
print(t)
t=[1,2,3]
print('And these are wrong:')
t.append([x]) # WRONG!
print(t)
t=[1,2,3]
t = t.append(x)# WRONG!
print(t)
t=[1,2,3]
t + [x] # WRONG!
print(t)
t=[1,2,3]
t = t + x # WRONG! -TypeError: can only concatenate list (not "int") to list
print(t)
t=[1,2,3]

These are correct:
[1, 2, 3, 4]
[1, 2, 3, 4]
[1, 2, 3, 4]
And these are wrong:
[1, 2, 3, [4]]
None
[1, 2, 3]


TypeError: can only concatenate list (not "int") to list

#### 3. Make copies to avoid aliasing.

If you want to use a method like sort that modifies the argument, but you need to keep the original list as well, you can make a copy.

### Gloassry

**list:** A sequence of values.


**element:** One of the values in a list (or other sequence), also called items.


**nested list:** A list that is an element of another list.


**accumulator:** A variable used in a loop to add up or accumulate a result.


**augmented assignment:** A statement that updates the value of a variable using an operator like +=.


**reduce:** A processing pattern that traverses a sequence and accumulates the elements into a single result.


**map:** A processing pattern that traverses a sequence and performs an operation on each element.


**filter:** A processing pattern that traverses a list and selects the elements that satisfy some criterion.


**object:** Something a variable can refer to. An object has a type and a value.


**equivalent:** Having the same value.


**identical:** Being the same object (which implies equivalence).


**reference:** The association between a variable and its value.


**aliasing:** A circumstance where two or more variables refer to the same object.


**delimiter:** A character or string used to indicate where a string should be split.

### Exercises

**Exercise 10.1.** Write a function called nested_sum that takes a list of lists of integers and adds up the elements from all of the nested lists. For example:

*t = [[1, 2], [3], [4, 5, 6]]*

*nested_sum(t)*

*\>>[21]*

In [35]:
def nested_sum(t):
    total = 0
    for nest in t:
        total += sum(nest)
    return total

t = [[1, 2], [3], [4, 5, 6]]

nested_sum(t)

21

**Exercise 10.2.** Write a function called cumsum that takes a list of numbers and returns the cumulative sum; that is, a new list where the ith element is the sum of the first i + 1 elements from the original list. For example:

*t = [1, 2, 3]*

*cumsum(t)*

*\>>[1, 3, 6]*

In [36]:
def cumsum(t):
    cumlist = t[:]
    for i in range(0,len(t)):
        if i == 0:
            cumlist[0] = t[0]
        else:
            cumlist[i] = t[i]+cumlist[i-1] 
    return cumlist

t = [1,2,3]
cumsum(t)
#print(t) check for modification

[1, 3, 6]

**Exercise 10.3.** Write a function called middle that takes a list and returns a new list that contains all but the first and last elements. For example:

*t = [1, 2, 3, 4]*

*middle(t)*

*\>>[2, 3]*

In [37]:
def middle(t):
    return t[1:-1]

t = [1,2,3,4]
middle(t)
#print(t) check for modification

[2, 3]

**Exercise 10.4.** Write a function called chop that takes a list, modifies it by removing the first and last elements, and returns None. For example:

*t = [1, 2, 3, 4]*

*chop(t)*

*\>>[2, 3]*

In [38]:
def chop(t):
    del t[0], t[-1]

t=[1,2,3,4]
chop(t)
t

[2, 3]

**Exercise 10.5.** Write a function called is_sorted that takes a list as a parameter and returns True if the list is sorted in ascending order and False otherwise. For example:

*is_sorted([1, 2, 2])*

True

*is_sorted(['b', 'a'])*

False

In [39]:
def is_sorted(t):
    return t == sorted(t)

print(is_sorted([1,2,2]))
is_sorted(['b','a'])

True


False

**Exercise 10.6.** Two words are anagrams if you can rearrange the letters from one to spell the other. Write a function called is_anagram that takes two strings and returns True if they are anagrams.

In [40]:
def is_anagrams(string1, string2):
    return sorted(list(string1)) == sorted(list(string2))

print(is_anagrams('swell', 'wells'))
is_anagrams('swell', 'well')

True


False

**Exercise 10.7.** Write a function called has_duplicates that takes a list and returns True if there is any element that appears more than once. It should not modify the original list.

In [41]:
def has_duplicates(t):
    return len(t) != len(set(t))

print(has_duplicates([1,2,3,4,5]))
print(has_duplicates([1,2,3,4,5,5,5,6]))
print(has_duplicates(['let', 'my', 'people', 'surf']))
print(has_duplicates(['let', 'my', 'people', 'surf', 'surf']))

False
True
False
True


**Exercise 10.8.** This exercise pertains to the so-called Birthday Paradox;


If there are 23 students in your class, what are the chances that two of you have the same birthday? You can estimate this probability by generating random samples of 23 birthdays and checking for matches. *Hint:* you can generate random birthdays with the randint function in the random module.

In [42]:
import random

def birthday_paradox_sim(students=23, experiments=1000):
    times_w_multiples = 0
    for i in range(0,experiments):
        birthdays = []
        for student in range(0,students):
            birthdays.append(random.randint(1,365))
        if has_duplicates(birthdays):
            times_w_multiples += 1
    return times_w_multiples/experiments

birthday_paradox_sim() #pretty good sim!

0.491

**Exercise 10.9.** Write a function that reads the file words.txt and builds a list with one element per word. Write two versions of this function, one using the append method and the other using the idiom t = t + [x]. Which one takes longer to run?

In [43]:
import time

def build_list():
    word_list = []
    fin = open('data/words.txt')
    for line in fin:
        word = line.strip()
        word_list.append(word)
    return sorted(word_list)

def build_list2():
    word_list = []
    fin = open('data/words.txt')
    for line in fin:
        word = line.strip()
        word_list = word_list + [word]
    return word_list

In [44]:
start_time = time.time()
t = build_list()
elapsed_time = time.time() - start_time

print(len(t))
print(t[:10])
print(elapsed_time, 'seconds')

113809
['aa', 'aah', 'aahed', 'aahing', 'aahs', 'aal', 'aalii', 'aaliis', 'aals', 'aardvark']
0.07283902168273926 seconds


In [45]:
start_time = time.time()
t = build_list2()
elapsed_time = time.time() - start_time

print(len(t))
print(t[:10])
print(elapsed_time, 'seconds')

113809
['aa', 'aah', 'aahed', 'aahing', 'aahs', 'aal', 'aalii', 'aaliis', 'aals', 'aardvark']
77.84410953521729 seconds


**Exercise 10.10.** To check whether a word is in the word list, you could use the in operator, but it would be slow because it searches through the words in order. Because the words are in alphabetical order, we can speed things up with a bisection search (also known as binary search), which is similar to what you do when you look a word up in the dictionary.


You start in the middle and check to see whether the word you are looking for comes before the word in the middle of the list. If so, you search the first half of the list the same way. Otherwise you search the second half.


Either way, you cut the remaining search space in half. If the word list has 113,809 words, it will take about 17 steps to find the word or conclude that it’s not there.


Write a function called in_bisect that takes a sorted list and a target value and returns True if the word is in the list and False if it’s not.


Or you could read the documentation of the bisect module and use that!

In [46]:
def bisect_search(t,word):
    if len(t) == 0:
        return False
    half = len(t)//2
    if half == 0 and t[half] != word:
        return False
    if word == t[half]:
        return True
    if word < t[half]:
        return bisect_search(t[:half], word)
    else:
        return bisect_search(t[half:],word)

In [47]:
word_list = build_list()

In [48]:
print(bisect_search(word_list, "kildler"))
print(bisect_search(word_list, "killer"))

False
True


**Exercise 10.11.** Two words are a “reverse pair” if each is the reverse of the other. Write a program that finds all the reverse pairs in the word list.

In [49]:
def is_reverse(word1, word2):
    if len(word1) != len(word2):
        return False
    i = 0
    j = len(word2)-1 #fixed
    while j >= 0: #fixed
        if word1[i] != word2[j]:
            return False
        i = i+1
        j = j-1
    return True

def reverse_pair(t):
    reverse_pairs = []
    for word in t:
        if bisect_search(reverse_pairs, word):
            pass
        if word == word[::-1]:
            pass
        elif bisect_search(t, word[::-1]):
            reverse_pairs.append(word)
            reverse_pairs.append(word[::-1])
    return reverse_pairs

In [50]:
#reverse_pairs = reverse_pair(word_list)

**Exercise 10.12.** Two words “interlock” if taking alternating letters from each forms a new word. For example, “shoe” and “cold” interlock to form “schooled”.

1. Write a program that finds all pairs of words that interlock. Hint: don’t enumerate all pairs!

2. Can you find any words that are three-way interlocked; that is, every third letter forms a word, starting from the first, second or third?

In [51]:
def interlock_words(t):
    interlocking_words = []
    for word in t:
        word1 = word[::2]
        word2 = word[1::2]
        if bisect_search(t, word1) and bisect_search(t, word2):
            interlocking_words.append((word,word1,word2))
    return interlocking_words

In [52]:
interlocking_words = interlock_words(word_list) #worked

In [53]:
print(len(interlocking_words))
interlocking_words[:10]

1254


[('aahs', 'ah', 'as'),
 ('abbey', 'aby', 'be'),
 ('abied', 'aid', 'be'),
 ('abies', 'ais', 'be'),
 ('abri', 'ar', 'bi'),
 ('abris', 'ars', 'bi'),
 ('abye', 'ay', 'be'),
 ('abyes', 'ays', 'be'),
 ('across', 'ars', 'cos'),
 ('added', 'add', 'de')]

In [54]:
def interlock_words_3way(t):
    interlocking_words = []
    for word in t:
        word1 = word[::2]
        word2 = word[1::2]
        word3 = word[2::2]
        if bisect_search(t, word1) and bisect_search(t, word2) and bisect_search(t, word3):
            interlocking_words.append((word,word1,word2,word3))
    return interlocking_words

In [55]:
threeway_interlocking_words = interlock_words_3way(word_list)

In [56]:
len(threeway_interlocking_words)
threeway_interlocking_words[:10]

[('abbey', 'aby', 'be', 'by'),
 ('abied', 'aid', 'be', 'id'),
 ('abies', 'ais', 'be', 'is'),
 ('adios', 'ais', 'do', 'is'),
 ('adnouns', 'anus', 'don', 'nus'),
 ('afreet', 'are', 'fet', 're'),
 ('afreets', 'ares', 'fet', 'res'),
 ('agios', 'ais', 'go', 'is'),
 ('agleam', 'ala', 'gem', 'la'),
 ('agreed', 'are', 'ged', 're')]